Make `LogLocator` respect `numticks`

dstansby opened the pull request:
## PR Summary
Fixes https://github.com/matplotlib/matplotlib/issues/11518. The logic for determining decades seems really odd, and dates back all the way to 2013 (https://github.com/matplotlib/matplotlib/pull/1686). Previously `numticks` wasn't respected; this should now be fixed, as verified by the changes to the tests I've made where `numticks=5`.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [n/a] New features are documented, with examples if plot related.
- [n/a] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [n/a] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

jklymak commented:
Is this the same as #20950 ?

dstansby commented:
> Is this the same as #20950 ?

Nope, that's a PR to do with contouring, not with the locator itself.

jklymak commented:
I took a quick look at this, but I can't really tell from the description or tests what has changed and if it is truly an improvement or not.  `len(ticks) < numticks` is the only test of the numticks that I can see, and that doesn't seem a very strict criteria.  I'm also leery of the "improves numerical issues when vmin and vmax are on a decade".  That should _always_ work, even if it means adding an epsilon to what is shown.

dstansby commented:
>`len(ticks) < numticks` is the only test of the numticks that I can see, and that doesn't seem a very strict criteria.

This isn't strict, but is what is promised in the docstring, and this PR fixes the locator to honour that promise.

> I'm also leery of the "improves numerical issues when vmin and vmax are on a decade". That should _always_ work, even if it means adding an epsilon to what is shown.

I've added some extra tests to check that it works for different bases, and removed that comment.

jklymak commented on lib/matplotlib/ticker.py:
```diff
@@ -2342,22 +2351,25 @@ def tick_values(self, vmin, vmax):
+            # If we have decided that the stride is bigger than the range, clip
+            # the stride back to the available range - 1 with a floor of 1.
+            # This prevents getting axis with only 1 tick visible.
+            stride = max(1, numdec)
```
code cov thinks this needs a test, and since its new, it is probably correct....

dstansby replied:
Thanks for spotting - it turns out this bit of code is redundant now and never hit, so I've removed it.

jklymak reviewed (commented):
I guess this needs an API change note to warn folks their plots will be different.  

I'm still a bit ignorant why this is a _better_ algorithm.  It seems to make sense, but given that it will break folks it needs a strong justification?

dstansby commented:
I've added an API change note. The tradeoff here is
Fixed bugs
- `maxticks` is respected
- The ticks on a log-scaled `errorbar` plot are now correct - see the third test image change.

Behaviour change
- In some cases (two of our test cases that use log scales, see 1st/2nd test image change) an extra tick is added to a log-axis, but there is still plenty of room to display the tick values so (beyond being a change) this isn't an issue.

jklymak commented:
I guess the image changes are my point though; many of them have twice as many ticks as before, and it is not clear to me why that is "right" or desirable...

jklymak commented:
The last axes in the last image test is what particularly jumps out at me as suddenly having a lot more ticks.  Maybe its fine though?

jklymak commented:
This needs a rebase...

Does anyone else have opinions about whether this is the correct algorithm or not?  Its a breaking change, but certainly numticks should be better respected for log scales!

dstansby commented:
Thanks - I'll wait on a third opinion before rebasing.

tacaswell commented:
We talked about this on the call and have a couple of quetions:

 - do we have a case where we actually return more ticks than we promise (at the "will be shown" level)
 - can we get away with being clearer in the docstring about what `numticks` means (because major ticks can only go on decades this ends up being way over-constrained)
 - We were not convinced that two of the images that got more ticks actually looked better, but this may just be an issue of moving across a hard edge (how many decades the major tick steps are) and there is no way around putting the jump _someplace_.  Is it possible to tweak that logic a bit to make it earlier again?

dstansby commented:
I've marked this as orphaned as I don't have the time/bandwidth to work more on this at the moment - if someone else wants to pick it up feel free.

dstansby commented:
I've put a large chunk of this in https://github.com/matplotlib/matplotlib/pull/24436, which is clearer and easier to review, so closing this.
